### PR TITLE
MAINT: review and refactor package

### PR DESCRIPTION
**Pull request by mmcky**
This is a refactor PR to make this extension more maintainable. 

It introduces:

1.  separate `exercise` and `solution` directive code (rather than inheriting from the same base) which lowers cognitive load when working with the objects. 
2. moves `env.exercise_list` to `env.sphinx_exercise_registry` as it was confusing to have `exercise` and `solution` nodes going into an object called `exercise_list`.
3. moves `post_transforms` into `post_transforms.py`
4. redesigns the `directives` and `post_transforms` by introducing `exercise-` and `solution-` titles and subtitles which enables targeted processes within the node (rather than have custom string modification code as before). This integrates more closely with sphinx and let's sphinx translators do the translation work from the `ast`
5. [latex] resolving hyperref links for `exercise` admonitions wasn't trivial as it requires recasting `numbered` references in a `post_transform` to a custom node to enable processing at the translator layer, as it wasn't possible to integrate with the base sphinx latex translator to include custom title text.
6. this PR introduces a change to how custom title text is formatted in admonition titles and links to admonitions

**Exercises:**
Screen Shot 2021-12-01 at 7 27 24 am

**Solutions:**
Screen Shot 2021-12-01 at 7 27 46 am

**Links to Solutions:**
 
Screen Shot 2021-12-01 at 7 28 53 am

### CODE REVIEW

**Comment by mmcky** [comment]
Fix the following test collections

- [ ] test_build
- [ ] test_solution
- [ ] test_latex
- [ ] test_reference

**Comment by mmcky** [comment]
- [ ] review docs
- [ ] review documentation (within code)
- [ ] review fixture diffs to ensure `references` are correct (as intended)

**Comment by mmcky** [comment]
@AakashGfude this is getting close. Integrating with the `sphinx` referencing systems has been a pain and it makes we wonder if we are doing this right. But I think there are not many better alternatives giving `xref` are resolved in `post_transforms` and they get applied at the document level rather than the project level. 

When you have time would you mind reviewing the `fixtures` and let me know what you think. The titles have changed a bit to add support for custom title text.

The main changes are I have worked directly with the `node` rather than carry out metadata etc. The `exercise` and `solution` nodes now contain `exercise-title` and `exercise-subtitle` nodes to enable identification etc. Anwyay open to thoughts and feedback on the code changes. 

If you could also take a look at `latex` -- I think `latex` support should now only be needed in the `node` visit and depart methods etc. I will look at `latex` next week if you don't have time.

**Comment by AakashGfude** [comment]
Pretty hefty PR 😬. Will carefully inspect and also add the changes you asked for.

**Comment by mmcky** [comment]
thanks @AakashGfude I should add that full `numfig` support isn't 100% but that is the way numfig is implemented so not sure what to do about that (other than take control and write numfig support into the extension for custom titling etc. 

I will also add some docs etc.

**Comment by choldgraf** [comment]
@mmcky are there particular questions that you have, where you'd like people to focus their reviews?

**Comment by mmcky** [comment]
thanks @choldgraf I was planning to get @AakashGfude to review (as a first round) then I will apply some final lessons learnt and update the docs. I was then planning to open up for wider review at that stage with some specific questions around `integration` with `references` in `sphinx`.

File: tests/test_exercise/_enum_numref_mathtitle.html
Comment (mmcky) [inline tests/test_exercise/_enum_numref_mathtitle.html:4]: @AakashGfude this is one of the `numref` fixtures I am referring to

File: sphinx_exercise/post_transforms.py
Comment (mmcky) [inline sphinx_exercise/post_transforms.py:27]: @AakashGfude perhaps this new `exercise_title` node is detached from the parent?

File: sphinx_exercise/post_transforms.py
Comment (mmcky) [inline sphinx_exercise/post_transforms.py:60]: @AakashGfude similarly here?

File: sphinx_exercise/post_transforms.py
Comment (AakashGfude) [reply sphinx_exercise/post_transforms.py:27]: @mmcky it does get attached to the `admonition node` later here: https://github.com/executablebooks/sphinx-exercise/blob/84a3d2ab9eb42600e3f5c1ccef267374ce386580/sphinx_exercise/post_transforms.py#L34

File: sphinx_exercise/post_transforms.py
Comment (AakashGfude) [reply sphinx_exercise/post_transforms.py:60]: similarly for this one.

File: sphinx_exercise/post_transforms.py
Comment (mmcky) [reply sphinx_exercise/post_transforms.py:27]: it replaces the child for sure -- but that isn't necessarily the same as

```
node += updated_title
```

Comment (mmcky) [reply sphinx_exercise/post_transforms.py:27]: I'll take a quick look

Comment (AakashGfude) [reply sphinx_exercise/post_transforms.py:27]: @mmcky adding visit, depart exercise_title support for latex here: https://github.com/executablebooks/sphinx-exercise/blob/84a3d2ab9eb42600e3f5c1ccef267374ce386580/sphinx_exercise/__init__.py#L179 , gets rid of the 
``
encountered title node not in section, topic, table, '
                              'admonition or sidebar
``
error, but needs some formatting for proper latex output.

Comment (mmcky) [reply sphinx_exercise/post_transforms.py:27]: Yeah it get's detached as the node metadata isn't retained 😮‍💨 

```
> sphinx-exercise/.tox/py38/lib/python3.8/site-packages/sphinx_exercise/post_transforms.py(25)resolve_enumerated_exercise_node_title()
-> title = node.children[0]
(Pdb) node.children[0]
<exercise_title: <#text: 'Exercise'><exercise_subtitle...>>
(Pdb) node.children[0].parent
<exercise_enumerable_node "exercise-1": <exercise_title...><section...>>
(Pdb) c

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB continue (IO-capturing resumed) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> sphinx-exercise/.tox/py38/lib/python3.8/site-packages/sphinx_exercise/post_transforms.py(39)resolve_enumerated_exercise_node_title()
-> return node
(Pdb) node.children[0].parent
```

Comment (mmcky) [reply sphinx_exercise/post_transforms.py:28]: should be fixed by ebf67f5

File: sphinx_exercise/post_transforms.py
Comment (mmcky) [reply sphinx_exercise/post_transforms.py:63]: > adding visit, depart exercise_title support for latex here:

@AakashGfude indeed. All the `LaTex` elements are current `raise NotImplementedError` for title and subtitle visit/depart.

File: sphinx_exercise/post_transforms.py
Comment (AakashGfude) [reply sphinx_exercise/post_transforms.py:28]: Thanks a lot @mmcky

**Comment by choldgraf** [comment]
@mmcky feel free to request a review once it's ready for others to look at it!

**Comment by mmcky** [comment]
@AakashGfude I did another review of the `post_transforms` for title and reference adjustments and simplified again.

https://github.com/executablebooks/sphinx-exercise/pull/37/commits/1665d17021bb7bd3ca692d857d6bd89485cf3c77

It would be great if you could have a look at the fixtures and let me know what you think (ie. they are resolving to good solutions in `html` and `xml`. The latex build is broken as a result of these changes.  Let me know what you think of these changes. Cheers.

**Comment by mmcky** [comment]
@AakashGfude I think this is looking pretty good. 

I have opened a test build here on a real-world test case

https://github.com/QuantEcon/continuous_time_mcs/pull/102

**Comment by mmcky** [comment]
- [x] fix LaTeX resolution of `\hyperref` for enumerated exercise nodes
- [ ] fix LaTeX resolution of `\hyperref` for solution nodes

**Comment by mmcky** [comment]
- [ ] tidy up the code and docs

**Comment by mmcky** [comment]
@AakashGfude I think we have now resolved the integration with `sphinx` references using `ref` and `numref` where we can. Let me know what you think of the changes now.

**Comment by AakashGfude** [comment]
@mmcky Looking good to me.

**Comment by mmcky** [comment]
@chrisjsewell @choldgraf we have given our best shot at integrating closely with sphinx in this PR. The focus of the re-write has been to alter the nodes in the `sphinx.ast` to remove custom code that was modifying strings, and using `ref` and `numref` when resolving cross-references between `exercise`, `solutions`, and `references` using `post_transforms`. We need to use `post_transforms` mainly as `pending_xref` nodes don't get updated until the `post_transform` stage with priority `10` -- what we also learnt is that `numref` is largely resolved at the `translator` layer (!)

We choose to integrate with the `std` domain as it allowed the use of `ref` and `numref` but that ended up being a lot of work to figure out how that cross-referencing infrastructure worked. I'd welcome any comments you have on this from your experience. I think what we have ended up with is a decent solution. There was one `workaround` for `numbered_references` for the `latex` builders due to a limitation in the default sphinx translator for latex. So we opted to use a `post_transform` to migrate `numbered_reference` to `exercise_latex_numbered_reference` nodes and add in our own simpler visit and depart method. I think this is a good solution pattern when the default just won't work.

**Review by AakashGfude — APPROVED** [review]

**Comment by AakashGfude** [comment]
@mmcky I think the pop_index is not required anymore https://github.com/executablebooks/sphinx-exercise/pull/37/files#diff-77185f1ed5e85402b80a05b1e8610ab9f33a9188e54393f8bdb3868594d0b5deR99

**Review by choldgraf — APPROVED** [review]
I took a quick pass through, with a focus on the docs since you mentioned that the main goal here was maintainability. I think in general it looks good to go (though didn't have time to do a deep dive since this changes a ton of things). The main comment I have is to make sure that the docstrings and documented well-enough that a person who isn't familiar with the codebase could orient themselves relatively quickly. There are likely a few places where there's assumptions being made and missing information because you all know a lot more about how the package works already :-)  it might also be helpful to have a short explainer of the structure of this repository, so that others know how it is implemented, where to look for things in the codebase etc. I don't think any of that needs to block this PR, but there could be some low-hanging fruit there if you want to give it a shot

File: sphinx_exercise/nodes.py
Comment (mmcky) [reply sphinx_exercise/nodes.py:7]: sure -- just didn't review that when we transferred to EBP :-). 👍

**Comment by mmcky** [comment]
> I took a quick pass through, with a focus on the docs since you mentioned that the main goal here was maintainability. I think in general it looks good to go (though didn't have time to do a deep dive since this changes a ton of things). The main comment I have is to make sure that the docstrings and documented well-enough that a person who isn't familiar with the codebase could orient themselves relatively quickly. There are likely a few places where there's assumptions being made and missing information because you all know a lot more about how the package works already :-) it might also be helpful to have a short explainer of the structure of this repository, so that others know how it is implemented, where to look for things in the codebase etc. I don't think any of that needs to block this PR, but there could be some low-hanging fruit there if you want to give it a shot

thanks for the review @choldgraf will do a final pass on docs and internal comments + docstrings today. I'l be adding a section with lessons learnt on the design of the package (which can serve as a source for a more general set of notes about writing a sphinx package for EBP)

**Comment by mmcky** [comment]
- [x] fixed issue with solution nodes to being clickable to `solution` nodes

**Comment by mmcky** [comment]
thanks @choldgraf and @AakashGfude I think this is good to be merged now. I'll merge tomorrow morning unless any further comments.

**Pull request merged**
